charms/tensorboard-controller/tests/integration/test_charm.py

```
@@ -0,0 +1,200 @@
+    else:
+        logger.info(
+            f"{resource_class_kind}/{resource_name} readyReplicas == {replicas} (waiting for '1')"
+        )
+
+    assert replicas == 1, f"Waited too long for {resource_class_kind}/{resource_name}!"
+
+
+async def test_create_tensorboard(ops_test: OpsTest):
+    """Test Tensorboard creation."""
+    lightkube_client = Client()
+    load_in_cluster_generic_resources(lightkube_client)
+
+    # Create PVC for Tensorboard logs and Tensorboard
+    resources = codecs.load_all_yaml(
+        "\n---\n".join([PVC_TEMPLATE_FILE.read_text(), TENSORBOARD_TEMPLATE_FILE.read_text()]),
+        context={"pvc_name": PVC_NAME, "tensorboard_name": TENSORBOARD_NAME},
+    )
+    for rsc in resources:
+        lightkube_client.create(rsc, namespace=ops_test.model_name)
+
+    try:
+        tensorboard_ready = lightkube_client.get(
+            TENSORBOARD_RESOURCE,
+            name=TENSORBOARD_NAME,
+            namespace=ops_test.model_name,
+        )
+    except ApiError:
+        assert False
+    assert tensorboard_ready, f"Tensorboard {ops_test.model_name}/{TENSORBOARD_NAME} not found!"
+
+    assert_replicas(lightkube_client, TENSORBOARD_RESOURCE, TENSORBOARD_NAME, ops_test.model_name)
+
+
+@pytest.mark.abort_on_fail
+async def test_remove_with_resources_present(ops_test: OpsTest):
+    """Test remove with all resources deployed.
+
+    Verify that all deployed resources that need to be removed are removed.
+    """
+    # remove deployed charm and verify that it is deleted
+    await ops_test.model.remove_application(app_name=APP_NAME, block_until_done=True)
+    assert APP_NAME not in ops_test.model.applications
+
+    # verify that all resources that were deployed are deleted
+    lightkube_client = Client()
+
+    # verify that all created CRDs and Services in namespace are deleted
+    for rsc in [
+        CustomResourceDefinition,
+        Service,
+    ]:
```

This test will fail fast (if CRD still exists, it'll never check for Service).  Is that the best way for this test to fail?  I'm thinking if I'm debugging this I could hit:
* run tests
* fail on CRD removal
* fix CRD removal
* fail on Service removal
* ...

What about checking all resources, then asserting they're all gone in a single assert?  That way someone might see that both CRDs and Services are left over at the same time.  

I haven't thought through how to make the failure message informative (doesn't help if at the end it just says "some resources left over - but you figure out which" :D ) but I think that's doable

---

I could keep stats and report them at the end of the test instead of asserting every single time. 

However, since the deletion itself is not fine-grained, I don't want to give the illusion of control over it.

---

yeah that's fair.  Go with your best judgement